### PR TITLE
feat: Add Traveling Salesman Problem algorithm and tests

### PR DESCRIPTION
**Pull request by harshithsaiv**
Implemented Traveling Salesman Problem using Dynamic Programming

### CODE REVIEW

**Comment by TruongNhanNguyen** [comment]
@harshithsaiv Sure, I can provide suggestions that you would like to apply into your code to improve it. I'm just a contributor to this repo and only have contribute rights (give suggestions to make the PR to be better, not the right to merge code into master branch).

**Comment by TruongNhanNguyen** [comment]
Although your implementation is LGTM, some points in your code need to be changed or improved further.
1. **Variable Names**: Variable names like `n`, `dp`, `mask`, `i`, and `j` are common in dynamic programming algorithms, but they could be more descriptive to enhance readability. For instance, `n` could be renamed to `numCities`, `dp` could be `costTable`, and so on.
2. We can refactor the code to separate the logic into reusable functions: 
   -   `initializeCostTable` initializes the cost table with maximum costs.
   -   `updateCostTable` updates the cost table with the minimum cost to reach each city.
   -   `findMinimumCost` finds the minimum cost to complete the traveling salesman route. 
   -   The main logic for solving the traveling salesman problem is encapsulated within the `TravelingSalesman` function, which orchestrates the use of the helper functions.

File: dynamic/travelingsalesman.go [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+	res := MAX
+	for i := 1; i < n; i++ {
+		if dp[allVisited][i]+cost[i][0] < res {
+			res = dp[allVisited][i] + cost[i][0]
+		}
+	}
+
+	return res
+}
```
Comment (TruongNhanNguyen) [inline dynamic/travelingsalesman.go:53]: Here are my suggestions for change
```suggestion
// initializeCostTable initializes a dynamic programming table with maximum costs.
func initializeCostTable(numCities int) [][]int {
	costTable := make([][]int, 1<<numCities)
	for i := range costTable {
		costTable[i] = make([]int, numCities)
		for j := range costTable[i] {
			costTable[i][j] = maxCost
		}
	}
	costTable[1][0] = 0
	return costTable
}

// updateCostTable updates the cost table with the minimum cost to reach each city.
func updateCostTable(costTable [][]int, mask, numCities int, cost [][]int) {
	for currentCity := 0; currentCity < numCities; currentCity++ {
		if (mask & (1 << currentCity)) != 0 {
			for nextCity := 0; nextCity < numCities; nextCity++ {
				if (mask & (1 << nextCity)) == 0 {
					newMask := mask | (1 << nextCity)
					if costTable[mask][currentCity]+cost[currentCity][nextCity] < costTable[newMask][nextCity] {
						costTable[newMask][nextCity] = costTable[mask][currentCity] + cost[currentCity][nextCity]
					}
				}
			}
		}
	}
}

// findMinimumCost finds the minimum cost to complete the traveling salesman route.
func findMinimumCost(costTable [][]int, allVisited, numCities int, cost [][]int) int {
	minCost := math.MaxInt32
	for i := 1; i < numCities; i++ {
		if costTable[allVisited][i]+cost[i][0] < minCost {
			minCost = costTable[allVisited][i] + cost[i][0]
		}
	}
	return minCost
}

// TravelingSalesman calculates the minimum cost to complete a round-trip through all cities,
// starting and ending at the first city, using dynamic programming with bitmasking to track visited cities.
func TravelingSalesman(cost [][]int) int {
	numCities := len(cost)
	allVisited := (1 << numCities) - 1
	costTable := initializeCostTable(numCities)

	for mask := 1; mask < (1 << numCities); mask += 2 {
		updateCostTable(costTable, mask, numCities, cost)
	}

	return findMinimumCost(costTable, allVisited, numCities, cost)
}

```

Comment (harshithsaiv) [reply dynamic/travelingsalesman.go:53]: Thank you for the review , I will make the necessary changes and make the code even more readable.

**Comment by TruongNhanNguyen** [comment]
LGTM.

**Review by TruongNhanNguyen — APPROVED** [review]

**Comment by jyzkm** [comment]
邮件已收到，祝好！

**Pull request closed without merging**